PR for homework e2e tests

Panda1941 opened the pull request:
Added e2e tests for most important functionalities of the website.

Panda1941 committed:
initial test structure

Panda1941 committed:
added more tests + adjusted folder structure

Panda1941 committed:
Added login and select flight page tests

Panda1941 committed:
Finished tests for all pages

Switched to not run in parallel (sometimes it caused weird errors on my machine, maybe its just a local error)
Some tests fail with reason, some fail quite randomly so i dont entirely know if I have made a mistake or its just weird behaviour 😃

Taura1 commented on tests/e2e/homework/tests/Login.spec.ts:
```diff
@@ -0,0 +1,79 @@
+    await loginPage.enterPassword(vars.correct_password);
+    await loginPage.submitSignIn();
+
+    await expect(page.locator('#flash_notice')).toBeVisible();
```
Ideally, in test file we should call majority of the functions - actions, assertions from page class. Please adjust throughout the solution

Taura1 replied:
Please extract common components, such as alerts and their validation to basePage class (then extend existing page classes with this new class).

Taura1 replied:
Please adjust throughout the whole solution

Taura1 commented on tests/e2e/homework/tests/Login.spec.ts:
```diff
@@ -0,0 +1,79 @@
+    await expect(page.locator('#flash_notice')).toBeVisible();
+});
+
+test.describe('Login with incorrect credentials', () => {
```
All of these tests have the same steps just different test data, please write it in such a way, that there is a single test code that executes different test data scenarios.

Taura1 commented on tests/e2e/homework/tests/PassengerDetails.spec.ts:
```diff
@@ -0,0 +1,46 @@
+    await passengerPage.enterPassengerSurname(vars.passenger_surname);
+    await passengerPage.submitPassengerDetails();
+
+    await expect(page.locator('h2')).toContainText('Pay by Credit Card');
```
this should go to page object - PaymentPage.ts
And it could be function named verifyPageHeader();

Taura1 commented on tests/e2e/homework/tests/PassengerDetails.spec.ts:
```diff
@@ -0,0 +1,46 @@
+    await expect(page.locator('h2')).toContainText('Pay by Credit Card');
+});
+
+test.describe('Enter invalid passenger details', () => {
```
Single test, multiple data sets -> go!

Taura1 commented on tests/e2e/homework/tests/Payment.spec.ts:
```diff
@@ -0,0 +1,74 @@
+    await expect(paymentPage.confirmation).toBeVisible();
+});
+
+test.describe('Enter invalid payment details', () => {
```
Single test, multiple data sets, where feasible

Taura1 commented on tests/e2e/homework/tests/SelectFlight.spec.ts:
```diff
@@ -0,0 +1,107 @@
+    await expect(selectFlightPage.selectFlightButton.first()).toBeChecked();
+
+    await selectFlightPage.continueToNextStep();
+    await expect(selectFlightPage.page.locator('h2')).toContainText('Passenger Details');
```
Move asserftion to page class

Taura1 commented on tests/e2e/homework/tests/SelectFlight.spec.ts:
```diff
@@ -0,0 +1,107 @@
+
+    await selectFlightPage.continueToNextStep();
+    await expect(selectFlightPage.page.locator('h2')).toContainText('Passenger Details');
+    await expect(selectFlightPage.page.locator('#container > div:nth-child(4) > b:nth-child(2)')).toContainText(vars.flight_from);
```
ugh, really ugly selectors :D find a way not to use full path, relative selectors from some elements should help. investigate ;)

Taura1 commented on tests/e2e/homework/tests/SelectFlight.spec.ts:
```diff
@@ -0,0 +1,107 @@
+
+        await expect(selectFlightPage.selectFrom).toHaveValue('');
+        await expect(selectFlightPage.continueButton).toBeDisabled();
+        // This throws an error, I believe the button should be disabled if the "from"airport is not selected.
```
either that, or error should appear on click

Taura1 commented on tests/e2e/homework/pages/Login.page.ts:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    async goto() {
+        await this.page.goto('https://travel.agileway.net/');
```
use baseURL config in playwright.config.ts

Panda1941 committed:
[WIP]

Panda1941 committed:
Tests adjusted to meet code review comments

Created base page class which contains some base locators and expected value checks.
Created parametrized tests as suggested.
Added base url to config.
Wasn't able to make trip selectors pretty so they are still ugly.
Assertions moved to page classes.

Taura1 commented on tests/e2e/homework/pages/BasePage.page.ts:
```diff
@@ -0,0 +1,39 @@
+        this.header = page.locator('h2');
+    }
+
+    async goto(){
```
Base page its not really a page you can go to. This function should not be here. This just a class that contains common methods, you don't have a page in the app that corresponds to it.

Taura1 commented on tests/e2e/homework/tests/Payment.spec.ts:
```diff
@@ -0,0 +1,74 @@
+        await paymentPage.submitPayment();
+
+        await expect(paymentPage.confirmation).toBeHidden();
+        // weirdly this fails
```
Investigate how DOM structure changes when the element is visible and when it's not. 
Adjust the assertion here accordingly ;)

Taura1 commented on tests/e2e/homework/pages/SelectFlight.page.ts:
```diff
@@ -0,0 +1,150 @@
+        this.continueButton = page.locator('[type="submit"]');
+        this.selectFlightButton = page.locator('[type="checkbox"]');
+
+        this.firstFromCity = page.locator('#container > div:nth-child(4) > b:nth-child(2)');
```
These locators are considered bad. A few changes to frontend (e.g. not bold but underlined text) and the test is broken, even though functionality of it did not change.
For the places you want to test - in best case scenario we'd ask devs to add test id on those elements.

Other example - we could take the full div element, it's text, and removing extra chars if any, and then asserting full string of text.

Taura1 commented on tests/e2e/homework/pages/Payment.page.ts:
```diff
@@ -0,0 +1,61 @@
+        this.expiryMonth = page.locator('[name="expiry_month"]');
+        this.expiryYear = page.locator('[name="expiry_year"]');
+        this.payNowButton = page.locator('[type="submit"]');
+        this.confirmation = page.locator('div#confirmation');
```
you don't need div here.
While writing selectors try to make them as simple, as short and as precice as possible

Taura1 commented:
With a few small tweaks on writing selectors, you're good to go for test automation contributions in a real project ;) 💯✅
